### PR TITLE
Add Parquet encoding statistics to column chunk metadata

### LINKED ISSUE #15313: [FEA] Add `encoding_stats` footer metadata to the parquet writer
**Is your feature request related to a problem? Please describe.**

The `parquet-cpp-arrow` writer includes ColumnChunk `encoding_stats` after the ColumnChunk statistics in the Parquet file footer. The encoding stats are useful for providing a total page count, tracking `RLE_DICTIONARY` fallback to `PLAIN` encoding, and verifying optional V2 encodings such as `DELTA_BYTE_ARRAY` and `DELTA_LENGTH_BYTE_ARRAY`.

[Parquet-tools ](https://pypi.org/project/parquet-tools/) is a simple command line interface to learn more about a parquet file.

Here is an example of the `encoding_stats` data from the writer `parquet-cpp-arrow version 14.0.2`
```
df = pd.DataFrame({'a': [1,2]})
df.to_parquet('cpp-arrow.pq')

df = cudf.DataFrame({'a': [1,2]})
df.to_parquet('cudf.pq')
```

`parquet-tools inspect --detail cpp-arrow.pq`

```
■■■■■■■■■■■■■■■■ColumnChunk
■■■■■■■■■■■■■■■■■■■■file_offset = 108
■■■■■■■■■■■■■■■■■■■■meta_data = ColumnMetaData
■■■■■■■■■■■■■■■■■■■■■■■■type = 2
■■■■■■■■■■■■■■■■■■■■■■■■encodings = list
■■■■■■■■■■■■■■■■■■■■■■■■■■■■0
■■■■■■■■■■■■■■■■■■■■■■■■■■■■3
■■■■■■■■■■■■■■■■■■■■■■■■■■■■8
■■■■■■■■■■■■■■■■■■■■■■■■path_in_schema = list
■■■■■■■■■■■■■■■■■■■■■■■■■■■■a
■■■■■■■■■■■■■■■■■■■■■■■■codec = 1
■■■■■■■■■■■■■■■■■■■■■■■■num_values = 2
■■■■■■■■■■■■■■■■■■■■■■■■total_uncompressed_size = 100
■■■■■■■■■■■■■■■■■■■■■■■■total_compressed_size = 104
■■■■■■■■■■■■■■■■■■■■■■■■data_page_offset = 36
■■■■■■■■■■■■■■■■■■■■■■■■dictionary_page_offset = 4
■■■■■■■■■■■■■■■■■■■■■■■■statistics = Statistics
■■■■■■■■■■■■■■■■■■■■■■■■■■■■max = b'\x02\x00\x00\x00\x00\x00\x00\x00'
■■■■■■■■■■■■■■■■■■■■■■■■■■■■min = b'\x01\x00\x00\x00\x00\x00\x00\x00'
■■■■■■■■■■■■■■■■■■■■■■■■■■■■max_value = b'\x02\x00\x00\x00\x00\x00\x00\x00'
■■■■■■■■■■■■■■■■■■■■■■■■■■■■min_value = b'\x01\x00\x00\x00\x00\x00\x00\x00'
■■■■■■■■■■■■■■■■■■■■■■■■encoding_stats = list
■■■■■■■■■■■■■■■■■■■■■■■■■■■■PageEncodingStats
■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■page_type = 2
■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■count = 1
■■■■■■■■■■■■■■■■■■■■■■■■■■■■PageEncodingStats
■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■encoding = 8
■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■count = 1
```

`parquet-tools inspect --detail cudf.pq`

```
■■■■■■■■■■■■■■■■ColumnChunk
■■■■■■■■■■■■■■■■■■■■meta_data = ColumnMetaData
■■■■■■■■■■■■■■■■■■■■■■■■type = 2
■■■■■■■■■■■■■■■■■■■■■■■■encodings = list
■■■■■■■■■■■■■■■■■■■■■■■■■■■■0
■■■■■■■■■■■■■■■■■■■■■■■■path_in_schema = list
■■■■■■■■■■■■■■■■■■■■■■■■■■■■a
■■■■■■■■■■■■■■■■■■■■■■■■codec = 1
■■■■■■■■■■■■■■■■■■■■■■■■num_values = 2
■■■■■■■■■■■■■■■■■■■■■■■■total_uncompressed_size = 33
■■■■■■■■■■■■■■■■■■■■■■■■total_compressed_size = 29
■■■■■■■■■■■■■■■■■■■■■■■■data_page_offset = 4
■■■■■■■■■■■■■■■■■■■■■■■■statistics = Statistics
■■■■■■■■■■■■■■■■■■■■■■■■■■■■max_value = b'\x02\x00\x00\x00\x00\x00\x00\x00'
■■■■■■■■■■■■■■■■■■■■■■■■■■■■min_value = b'\x01\x00\x00\x00\x00\x00\x00\x00'
```

### PR DESCRIPTION
**Pull request by etseidl**
## Description
Closes #15313

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/compact_protocol_reader.cpp [hunk 1]
```diff
@@ -16,6 +16,9 @@
 
 #include "compact_protocol_reader.hpp"
 
+#include "parquet.hpp"
```
Comment (ttnghia) [inline cpp/src/io/parquet/compact_protocol_reader.cpp:19]: Please remove the empty line above :smile:

File: cpp/src/io/parquet/compact_protocol_writer.cpp [hunk 2]
```diff
@@ -16,6 +16,8 @@
 
 #include "compact_protocol_writer.hpp"
 
+#include "parquet.hpp"
```
Comment (ttnghia) [inline cpp/src/io/parquet/compact_protocol_writer.cpp:19]: Similar to above.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/parquet/compact_protocol_reader.cpp (hunk 1 above)
Comment (ttnghia) [reply cpp/src/io/parquet/compact_protocol_reader.cpp:19]: It's weird. Definitely clang-format bug :sweat:

Comment (etseidl) [reply cpp/src/io/parquet/compact_protocol_reader.cpp:19]: Yeah. Maybe a special group for header with the same root as the compilation unit 🤷

**Review by mhaseeb123 — APPROVED** [review]

**Comment by mhaseeb123** [comment]
Thanks for the effort Ed, looks good to me!

File: cpp/src/io/parquet/compact_protocol_reader.cpp (hunk 1 above)
Comment (ttnghia) [reply cpp/src/io/parquet/compact_protocol_reader.cpp:19]: @harrism Would you mind looking into this please?

**Pull request merged**